### DISCUSSION TITLE
[2023-12-12] Team GUI Demo Day

### DISCUSSION THREAD

**Discussion by farmaazon** [enso-org/enso#8526] [Weekly Demo Day]
* The main focus were #8462 (finished, but still facing some CI issues)
  * The build script `./run gui2 test` - runs both unit and e2e tests in console
  * For debugging, one can run `npm --workspace=enso-gui2 run test:e2e -- --ui` with nice interface
  * There is also VSCode plugin, but I haven't used it yet.

![image](https://github.com/enso-org/enso/assets/3919101/1a1d7d4a-8a0b-446c-9710-d250627298ee)

Besides, fixed several discovered regressions:
* Worked around an error in production builds preventing opening projects #8458
* The not working visualizations #8479 
* Editing by click: #8456 
* #8507 - waiting for review  :
  * Delete key not working in CB and in Code Editor
  * Cursor not placed at the end when adding new node with source node
  * Garbage added when edited node while having another one selected.
  * Closing CB when navigating.
  * Too eager selecting component after filtering update.
  * Premature node creation when dropping edges
  * Discarding changes when clicking-off CB

**Comment by MichaelMauderer** [comment]
Implemented entering/leaving nodes (#8492) + integration with breadcrumbs (#7788): 


[Peek 2023-12-07 13-28.webm](https://github.com/enso-org/enso/assets/1428930/85da0dd6-0c08-4bef-b8e1-ba9573a3e6cd)


[Peek 2023-12-12 10-46.webm](https://github.com/enso-org/enso/assets/1428930/b4f76193-b8b7-4b78-ba11-89654a8c1505)

**Comment by vitvakatu** [comment]
- Fixing a couple of minor issues in dynamic dropdowns: https://github.com/enso-org/enso/pull/8478
- A small fix for the old GUI: https://github.com/enso-org/enso/pull/8464
- Working on https://github.com/enso-org/enso/issues/8067 – waiting for AST edits.
- Fixing issues in the documentation panel: https://github.com/enso-org/enso/issues/8135


https://github.com/enso-org/enso/assets/6566674/7a701fae-4338-4383-b11b-d0e435f1251c


https://github.com/enso-org/enso/assets/6566674/2f609ff6-86bf-47e1-9737-649e72f94893

**Comment by kazcw** [comment]
This past week:
- Ported new app features to the new AST, and merged the AST PR: #8263
- Added a new syntax in the parser (#8470) and ensured support for an upcoming feature (`default`) works consistently between frontend/backend/parser.
- Implemented simpler AST match/replace based on new AST edit APIs, to support #8072. [PR 1/2 for #8367; will be ready this morning.]

Next:
- Part 2/2 PR for #8367. Will eliminate the need to maintain an IdMap, and will make reordering blocks simple (see #8494).

**Comment by somebody1234** [comment]
Mostly various small fixes/other small QoL improvements this week.

Merged:
- Minor QoL improvement: #8506
- Minor fix: #8488
- Temporary workaround: #8461
  - Long term plans according to the issue's comments is that `PropertyAccess` shouldn't contain nested `Ident` *nodes*, only nested `Ident` *tokens*
- Implemented old feature request: #8490
- Fix on Dashboard side: #8445

Worked on:
- Minor restructure of `@/util` #8501
- Porting to new AST: #8440
- Fixes for downloading Enso Font in CI: #8508
- (Heavily WIP) Fixing crashes when calling `openFile` #8514

**Comment by mwu-tow** [comment]
I've done an extensive investigation of the NSIS installer issue.

Essentially, the issues are caused by the paths exceeding `MAX_PATH`, NSIS has trouble properly dealing with them. Also, diagnostics is really confusing, as the error handling code (and messages) are unclear or plain wrong.

The issue happens only on newer electron-builder versions, because a new update procedure was provided that first tries to backup the original installation by moving it away to a temporary location. This temporary location has a longer path (`C:\Users\mwurb\AppData\Local\Temp\nsz758E.tmp\old-install` vs `C:\Users\mwurb\AppData\Local\Programs\Enso`) which causes some paths to go over the limit.

However, we are already balancing on the limit. If we have a slightly longer path, the issue will also appear during initial installation. Whether we can reasonably avoid the problem on our side is being discussed in https://github.com/orgs/enso-org/discussions/8509

***

The alternative tools offered by electron-builder (and electron-forge) are not really suitable for us, as they either have similar issues or are not really maintained.

***

As an alternative approach, I'm investigating how feasible it would be to provide our own installation script. We already skip the vast majority of NSIS features (as we want one-click installation) and do not care about its other characteristics (like being only 35kB or supporting older Windows).

**Comment by Frizi** [comment]
- Finishing work on setting values for placeholder arguments (and unifying widget value updates in general). This is somewhat broken due to not having AST updates implemented yet.  #8258 
- Addressed incorrect whitespace rendering between widgets. Now spaces around operators will show up properly.

https://github.com/enso-org/enso/assets/919491/362dae59-cccb-4d89-b1f6-6ad6ffcbe2cd

<img width="399" alt="image" src="https://github.com/enso-org/enso/assets/919491/9dd137b7-1719-4ef6-82fe-d66192f97831">

**Comment by JaroslavTulach** [comment]
### 2-3x test speedup by reduced memory usage

- #8418

#### Initially
![Screenshot from 2023-11-29 11-06-04](https://github.com/enso-org/enso/assets/292128/af437d8a-9111-4bd6-9033-a59030c7ebed)


#### Now
![Screenshot from 2023-11-29 15-57-34](https://github.com/enso-org/enso/assets/292128/3c1f8aef-fe7e-4f5b-a236-12c86ea8b906)


### Inline Execution Work

...continues...

### Performance of the `EnsoHashMap` shall no longer be a problem with #8425 

![obrazek](https://github.com/enso-org/enso/assets/26887752/81d39701-aa31-41ba-8246-efcc3a21c516)

`org_enso_benchmarks_generated_Enso_Hash_Map_100000_Enso_Incremental` achieves `4.06ms` per operation which makes it **slightly faster** than `org_enso_benchmarks_generated_Enso_Hash_Map_100000_Java_Incremental` benchmark.

![obrazek](https://github.com/enso-org/enso/assets/26887752/bd09ee10-e167-4b42-be3b-1b3756c619db)

`org_enso_benchmarks_generated_Enso_Hash_Map_100000_Enso_Replacement` has been sped up **400 times** and is now just 50% slower than the `org_enso_benchmarks_generated_Enso_Hash_Map_100000_Java_Incremental` - that's a good result given that the Java `HashMap` doesn't need to keep the referential transparency at all.
